src/reanimated2/animation/clamp.ts

```diff
@@ -0,0 +1,102 @@
+'use strict';
+import { defineAnimation, getReduceMotionForAnimation } from './util';
+import type {
+  Animation,
+  Timestamp,
+  AnimatableValue,
+  AnimationObject,
+  ReduceMotion,
+} from '../commonTypes';
+import type { ClampAnimation } from './commonTypes';
+
+type withClampType = <T extends AnimatableValue>(
+  config: { min?: number; max?: number; reduceMotion?: ReduceMotion },
+  clampedAnimation: T
+) => T;
+
+// TODO This feature is not documented yet
+export const withClamp = function <T extends AnimationObject<number>>(
+  config: { min?: number; max?: number; reduceMotion?: ReduceMotion },
+  _clampedAnimation: T | (() => T)
+): Animation<ClampAnimation> {
+  'worklet';
+  return defineAnimation<ClampAnimation, T>(
+    _clampedAnimation,
+    (): ClampAnimation => {
+      'worklet';
+      const clampedAnimation =
+        typeof _clampedAnimation === 'function'
+          ? _clampedAnimation()
+          : _clampedAnimation;
+
+      function clampOnFrame(
+        animation: ClampAnimation,
+        now: Timestamp
+      ): boolean {
+        const finished = clampedAnimation.onFrame(clampedAnimation, now);
+
+        if (clampedAnimation.current === undefined) {
+          // This should never happen
+          // TODO check if nextAnimation.current should be optional
+          console.warn(
+            "[Reanimated] Error inside 'withClamp' animation, the inner animation has invalid current value"
+          );
+          return true;
+        } else {
+          const clampUpper = config.max ? config.max : clampedAnimation.current;
+          const clampLower = config.min ? config.min : clampedAnimation.current;
+          animation.current = Math.max(
+            clampLower,
+            Math.min(clampUpper, clampedAnimation.current)
+          );
```

Comment:
Unfortunately it doesn't work for some edge cases:
<img width="216" alt="image" src="https://github.com/software-mansion/react-native-reanimated/assets/36106620/9b1443b3-1667-4a5a-8081-4201cec90915">
I think that the better option will be just it:
```suggestion
if (config.max < clampedAnimation.current) {
  animation.current = config.max;
}
if (config.min > clampedAnimation.current) {
  animation.current = config.min;
}
```